### ISSUE TITLE
Question about calculation of Q and transpose(K).

### ISSUE THREAD

**Issue by jaes77** [lucidrains/flash-attention-jax#10]
Thanks for your effort to make this great platform.

In normal attention, the input of softmax function is a form of matmul(Q,K_T) and its dimension is (batch, num_heads, q_len, k_len)
Also, the attention mask is like a trigonal shape (total shape is could be q_len x k_len)
so, matmul(q, k_t) is masked with the attention mask.

However, I don't understand how matmul(q_chunk, transposed k_chunk) works and results in masked input of softmax compared with original attention algorithm flow at the code lines below.

https://github.com/lucidrains/flash-attention-jax/blob/5727815b12aa4c01059b82020c891842d719ff46/flash_attention_jax/flash_attention.py#L34-L37

Can you explain it with details?
